sql-plugin/src/main/scala/org/apache/spark/sql/rapids/RapidsShuffleInternalManagerBase.scala

```diff
@@ -275,6 +275,28 @@ abstract class RapidsShuffleThreadedWriterBase[K, V](
   val diskBlockObjectWriters = new mutable.HashMap[Int, (Int, DiskBlockObjectWriter)]()
 
   override def write(records: Iterator[Product2[K, V]]): Unit = {
+    // Iterating the `records` may involve some heavy computations.
+    // This iterator is used to track how much time we spend for such computations.
+    class TimeTrackingIterator extends Iterator[Product2[K, V]] {
+      var iterateTimeNs: Long = 0L
+
+      override def hasNext: Boolean = {
+        val start = System.nanoTime()
+        val ret = records.hasNext
```

Comment:
could we have `TimeTrackingIterator` take the iterator to wrap? Lets have it also defined outside of the `write` method.

Reply:
Sure, where is the best place to have it defined?

Reply:
I considered `GpuExec.scala` before, but there is already a similar class, `CollectTimeIterator`, which wraps the iterator with an nvtx range. Do we want to keep these iterators of `CollectTimeIterator` and `TimeTrackingIterator` separate?

Reply:
I just meant in this scala file, just something that clearly took the iterator. The reason I prefer that is that this way I don't have an iterator going into `write` and then another iterator that is also able to advance it. It is clear that you are supposed to use the wrapper. It is a bit of a nit.

Reply:
Ah I see. This makes sense. I will update the code soon.

Reply:
Ok, added another `write` function that takes a `TimeTrackingIterator`, so that the `records` iterator wrapped in it cannot be accessed from the outside. This new function is called from the original function.